weave/upload.py

```diff
@@ -321,85 +330,70 @@ def setup_temp_dir_for_staging_prior_to_fs(self):
-                            else:
-                                self.file_system.upload(local_path,
-                                                        file_upload_path)
-                        else:
-                            fid["stub"] = True
-                            fid["upload_path"] = "stub"
-                        supplement_data["integrity_data"].append(fid)
+                        supplement_data["integrity_data"].append(file_int_dat)
             else:
-                fid = derive_integrity_data(
-                    str(item_path),
-                    file_system=self.file_system,
-                    source_file_system=self.source_file_system)
-                if upload_item["stub"] is False:
-                    fid["stub"] = False
-                    file_upload_path = os.path.join(
-                        self.kwargs.get("upload_directory"),
-                        os.path.basename(item_path)
-                    )
-                    fid["upload_path"] = str(file_upload_path)
-                    base_path = os.path.split(file_upload_path)[0]
-                    if not self.file_system.exists(base_path):
-                        self.file_system.mkdir(base_path)
-                    if self.source_file_system == self.file_system:
-                        self.file_system.copy(str(item_path),
-                                              file_upload_path)
-                    elif isinstance(self.source_file_system,s3fs.S3FileSystem):
-                        self.source_file_system.get(str(item_path),
-                                                    file_upload_path)
-                    else:
-                        self.file_system.upload(str(item_path),
-                                            file_upload_path)
-                else:
-                    fid["stub"] = True
-                    fid["upload_path"] = "stub"
-                supplement_data["integrity_data"].append(fid)
+                file_int_dat = self.handle_file_integrity_data(
+                    str(item_path), upload_item, item_path
+                )
+                supplement_data["integrity_data"].append(file_int_dat)
         self.kwargs["supplement_data"] = supplement_data
 
+    def handle_file_integrity_data(self, local_path, upload_item, item_path):
+        """Gathers the file integrity data, handles stub logic"""
+        file_int_dat = derive_integrity_data(
+            str(local_path),
+            file_system=self.file_system,
+            source_file_system=self.source_file_system,
+        )
+        if upload_item["stub"] is False:
+            file_int_dat["stub"] = False
+            file_upload_path = self.construct_file_upload_path(
+                local_path, item_path
+            )
+            file_int_dat["upload_path"] = str(file_upload_path)
+            self.handle_file_upload(local_path, file_upload_path)
+        else:
+            file_int_dat["stub"] = True
```

Comment:
this line and line 364 could be condensed down to file_int_dat["stub"] = upload_item["stub"], but I'm not going to block because of it.